File: mcp-code-intelligence-python/src/mcp_code_intel/analyzers/graph_analysis/module_summarizer.py
"""KSA-163: Module Summarizer."""
from __future__ import annotations
import sqlite3
from . import ModuleSummary
from .utils import GraphLoader
from .circular_dep_detector import CircularDepDetector
from .hot_path_analyzer import HotPathAnalyzer
from .dead_import_detector import DeadImportDetector


class ModuleSummarizer:
    def __init__(self, conn: sqlite3.Connection) -> None:
        self._conn = conn
        self._gl = GraphLoader(conn)

    def summarize(self, module_name: str | None = None) -> list[ModuleSummary]:
        modules = self._get_modules(module_name)
        results = []
        for name, fc, sc in modules:
            cd = CircularDepDetector(self._gl).detect(module=name)
            hp = HotPathAnalyzer(self._gl).analyze(module=name, limit=5)
            di = DeadImportDetector(self._conn).detect(module=name)
            avg = self._avg_complexity(name)
            results.append(ModuleSummary(name, fc, sc, len(cd), hp, len(di), avg))
        return results

    def _get_modules(self, name: str | None) -> list[tuple[str, int, int]]:
        sql = "SELECT name, file_count, symbol_count FROM modules"
        params: list = []
        if name: sql += " WHERE name = ?"; params.append(name)
        return self._conn.execute(sql, params).fetchall()

    def _avg_complexity(self, module: str) -> float | None:
        row = self._conn.execute(
            "SELECT AVG(c.cyclomatic_complexity) FROM complexity c "
            "JOIN symbols s ON s.id = c.symbol_id JOIN files f ON f.id = s.file_id "
            "WHERE f.module = ?", (module,),
        ).fetchone()
        return row[0] if row and row[0] is not None else None

File: backend/src/engine/analyzers/graph-analysis/ModuleSummarizer.ts
/**
 * KSA-163: Module Summarizer — Aggregates quality metrics per module.
 */

import Database from 'better-sqlite3';
import type { ModuleSummary } from './types.js';
import { GraphLoader } from './utils/GraphLoader.js';
import { CircularDepDetector } from './CircularDepDetector.js';
import { HotPathAnalyzer } from './HotPathAnalyzer.js';
import { DeadImportDetector } from './DeadImportDetector.js';

export class ModuleSummarizer {
  private db: Database.Database;
  private graphLoader: GraphLoader;

  constructor(db: Database.Database) {
    this.db = db;
    this.graphLoader = new GraphLoader(db);
  }

  /** Generate summary for a specific module or all modules. */
  summarize(moduleName?: string): ModuleSummary[] {
    const modules = this.getModules(moduleName);
    const results: ModuleSummary[] = [];

    for (const mod of modules) {
      const circularDetector = new CircularDepDetector(this.graphLoader);
      const hotPathAnalyzer = new HotPathAnalyzer(this.graphLoader);
      const deadImportDetector = new DeadImportDetector(this.db);

      const circularDeps = circularDetector.detect({ module: mod.name });
      const hotPaths = hotPathAnalyzer.analyze({ module: mod.name, limit: 5 });
      const deadImports = deadImportDetector.detect({ module: mod.name });
      const avgComplexity = this.getAvgComplexity(mod.name);

      results.push({
        module: mod.name,
        fileCount: mod.fileCount,
        symbolCount: mod.symbolCount,
        circularDeps: circularDeps.length,
        hotPaths,
        deadImports: deadImports.length,
        avgComplexity,
      });
    }

    return results;
  }

  private getModules(name?: string): Array<{ name: string; fileCount: number; symbolCount: number }> {
    let sql = 'SELECT name, file_count as fileCount, symbol_count as symbolCount FROM modules';
    const params: unknown[] = [];
    if (name) {
      sql += ' WHERE name = ?';
      params.push(name);
    }
    return this.db.prepare(sql).all(...params) as Array<{
      name: string; fileCount: number; symbolCount: number;
    }>;
  }

  private getAvgComplexity(module: string): number | null {
    const row = this.db.prepare(`
      SELECT AVG(c.cyclomatic_complexity) as avg
      FROM complexity c
      JOIN symbols s ON s.id = c.symbol_id
      JOIN files f ON f.id = s.file_id
      WHERE f.module = ?
    `).get(module) as { avg: number | null } | undefined;
    return row?.avg ?? null;
  }
}
